tamara_gadieva/graph.cpp

```diff
@@ -42,34 +38,73 @@ void Vertex::add_edge_id(const EdgeId& id) {
   edge_ids_.push_back(id);
 }
 
-VertexId Graph::add_vertex(const int depth) {
+void Graph::fill_vertices_ids_in_depth(int depth, VertexId id) {
+  if (depth >= vertices_ids_in_depth_.size())
+    vertices_ids_in_depth_.resize(depth + 1);
+  vertices_ids_in_depth_[depth].push_back(id);
+  if (id != 0) {
+    for (auto vertex_id_ind = vertices_ids_in_depth_[0].begin();
+         vertex_id_ind < vertices_ids_in_depth_[0].end(); vertex_id_ind++) {
+      if (*vertex_id_ind == id)
+        vertices_ids_in_depth_[0].erase(vertex_id_ind);
+    }
+  }
+}
+
+VertexId Graph::add_vertex() {
   auto new_vertex_id = get_new_vertex_id();
-  vertices_.emplace_back(new_vertex_id, depth);
+  vertices_.emplace_back(new_vertex_id);
+  fill_vertices_ids_in_depth(0, new_vertex_id);
   return new_vertex_id;
 }
 
+std::vector<EdgeId> Graph::get_edge_ids(const VertexId& id) {
+  const auto& vertex = get_vertex(id);
+  return vertex.get_edge_ids();
+}
+
+Edge::Color Graph::define_edge_color(const VertexId& from_vertex_id,
+                                     const VertexId& to_vertex_id) {
+  auto& from_vertex = get_vertex(from_vertex_id);
+  auto& to_vertex = get_vertex(to_vertex_id);
+  if (get_edge_ids(to_vertex_id).size() == 0)
+    return Edge::Color::Gray;
+  if (from_vertex_id == to_vertex_id)
+    return Edge::Color::Green;
+  if (to_vertex.depth - from_vertex.depth == 1 &&
+      !is_connected(from_vertex_id, to_vertex_id))
+    return Edge::Color::Yellow;
+  if (to_vertex.depth - from_vertex.depth == 2)
+    return Edge::Color::Red;
+  throw std::runtime_error("Unreachable code");
+}
+
 void Graph::add_edge(const VertexId& from_vertex_id,
-                     const VertexId& to_vertex_id,
-                     const Edge::Color& edge_color) {
+                     const VertexId& to_vertex_id) {
   assert(has_vertex(from_vertex_id) && "Vertex doesn't exist");
   assert(has_vertex(to_vertex_id) && "Vertex doesn't exist");
-  if (edge_color != Edge::Color::Green)
-    assert(!is_connected(from_vertex_id, to_vertex_id) &&
-           "Vertices are already connected");
-
-  const auto& new_edge = edges_.emplace_back(get_new_edge_id(), from_vertex_id,
-                                             to_vertex_id, edge_color);
+  assert(!is_connected(from_vertex_id, to_vertex_id) &&
+         "Vertices are already connected");
   auto& from_vertex = get_vertex(from_vertex_id);
   auto& to_vertex = get_vertex(to_vertex_id);
+  auto edge_color = define_edge_color(from_vertex_id, to_vertex_id);
+  const auto& new_edge = edges_.emplace_back(get_new_edge_id(), from_vertex_id,
+                                             to_vertex_id, edge_color);
   from_vertex.add_edge_id(new_edge.id);
   if (edge_color != Edge::Color::Green)
     to_vertex.add_edge_id(new_edge.id);
+
+  if (edge_color == Edge::Color::Gray)
+    to_vertex.depth = from_vertex.depth + 1;
+  fill_vertices_ids_in_depth(to_vertex.depth, to_vertex.id);
 }
```

Comment:
1) Глубины должны обновляться только для  серых граней, наверное ошиблась, потому что скобочки не ставишь у `if` условий
2) Скорее не "заполнить на глубине", а "обновить глубину вершины"
```suggestion
  if (edge_color == Edge::Color::Gray) {
    const int new_depth = from_vertex.depth + 1;
    update_vertex_depth(to_vertex.id, new_depth);
  }
```
Соответственно функция и обновляешь глубину у вершины, и обновляет карту глубин. Будет более связанная логика, чем у тебя сейчас.